server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DKIMVerify.java

```diff
@@ -0,0 +1,112 @@
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.List;
+
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+
+import org.apache.james.jdkim.DKIMVerifier;
+import org.apache.james.jdkim.api.BodyHasher;
+import org.apache.james.jdkim.api.Headers;
+import org.apache.james.jdkim.api.SignatureRecord;
+import org.apache.james.jdkim.exceptions.FailException;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMailet;
+
+import org.apache.james.transport.mailets.jdkim.MimeMessageHeaders;
+import org.apache.james.transport.mailets.jdkim.CRLFOutputStream;
+import org.apache.james.transport.mailets.jdkim.HeaderSkippingOutputStream;
+
+/**
+ * This mailet verify a message using the DKIM protocol
+ * 
+ * Sample configuration:
+ * <pre><code>
+ * &lt;mailet match=&quot;All&quot; class=&quot;DKIMVerify&quot;&gt;
+ * &lt;/mailet&gt;
+ * </code></pre>
+ * 
+ * By default the mailet assume that Javamail will use LF instead of CRLF 
+ * so it will verify the hash using converted newlines. If you don't want this 
+ * behaviout then set forceCRLF attribute to false. 
+ */
+public class DKIMVerify extends GenericMailet {
+
+    public static final String DKIM_AUTH_RESULT_ATTRIBUTE = "jDKIM.AUTHRESULT";
+    
+    protected DKIMVerifier verifier = null;
+    private boolean forceCRLF;
+
+    @Override
+    public void init() throws MessagingException {
+        verifier = new DKIMVerifier();
+        forceCRLF = getInitParameter("forceCRLF", true);
+    }
+    
+    public void service(Mail mail) throws MessagingException {
+        try {
+            MimeMessage message = mail.getMessage();
+            List<SignatureRecord> res = verify(verifier, message, forceCRLF);
+            if (res == null || res.isEmpty()) {
+                // neutral
+                mail.setAttribute(DKIM_AUTH_RESULT_ATTRIBUTE, "neutral (no signatures)");
+            } else {
+                // pass
+                StringBuilder msg = new StringBuilder();
+                msg.append("pass");
+                for (SignatureRecord rec : res) {

```

Comment:
Moving is one issue, refactoring is an other. Do we want to take time doing this?

(It is an external contribution, we can ask Shrumit to do it...)